src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -60,28 +118,315 @@ def validate(self) -> None:
+        3. Create a new list of records to be inserted which excludes the
+        records which are present in the above two lists.
+        4. Insert the new list of records in a transaction.
+
+        Once the transaction succeeds, we update the key_results by
+        performing a read of the organization_id and string which were
+        supposed to be originally inserted.
+        """
+
+        def insert_rw_transaction_uow(
+            transaction: Any, rows: Sequence[SpannerIndexerModel]
+        ) -> None:
+            existing_org_string_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.organization_id, row.string] for row in rows]),
+                index=self._get_unique_org_string_index_name(use_case_id),
+            )
+
+            existing_id_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.id] for row in rows]),
+            )
+
+            existing_records = list(existing_org_string_results) + list(existing_id_results)
+            exisiting_org_string_set = {(row[0], row[1]) for row in existing_records}
+            existing_id_string_set = {row[2] for row in existing_records}
+
+            missing_records = []
+            for row in rows:
+                if (
+                    row.organization_id,
+                    row.string,
+                ) not in exisiting_org_string_set and row.id not in existing_id_string_set:
+                    missing_records.append(row)
+
+            if missing_records:
+                transaction.insert(
+                    self._get_table_name(use_case_id),
+                    columns=get_column_names(),
+                    values=missing_records,
+                )
+
+        metrics.incr(
+            _INDEXER_DB_INSERT_METRIC,
+            tags={"batch": "false"},
+        )
+        transaction_succeeded = False
+        while not transaction_succeeded:
+            try:
+                self.database.run_in_transaction(insert_rw_transaction_uow, rows_to_insert)
+                transaction_succeeded = True
+            except google.api_core.exceptions.AlreadyExists:
+                metrics.incr(_INDEXER_DB_RW_TRANSACTION_FAILED_METRIC)
+            else:
+                with self.database.snapshot() as snapshot:
+                    result = snapshot.read(
+                        table=self._get_table_name(use_case_id),
+                        columns=["organization_id", "string", "id"],
+                        keyset=spanner.KeySet(
+                            keys=[[row.organization_id, row.string] for row in rows_to_insert]
+                        ),
+                        index=self._get_unique_org_string_index_name(use_case_id),
+                    )
```

Comment:
I suspect this may not work as this will be a stale read. 
Though can't we return the results directly from the transaction code? If the transaction is successful we know what we found and what we added. 
If the transaction function cannot return anything, can't you pass it a mutable object to fill in ? So we avoid another read

Reply:
This would perform a strong read. We do not provide staleness parameter during this read operation https://cloud.google.com/spanner/docs/reads#perform-stale-read.

I was thinking of filling in the keyresults from the transaction code. But I leaned towards doing an additional read in order to keep the transaction code simpler. I will update the PR.

Reply:
Updated the PR such that if transaction succeeds, the key results are updated based on the records which were written successfully.